File: app/src/main/java/org/metabrainz/mobile/data/sources/api/LoginService.java
package org.metabrainz.mobile.data.sources.api;

import org.metabrainz.mobile.data.sources.api.entities.AccessToken;
import org.metabrainz.mobile.data.sources.api.entities.userdata.UserInfo;

import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Url;

public interface LoginService {

    @FormUrlEncoded
    @POST
    Call<AccessToken> getAccessToken(@Url String url,
                                     @Field("code") String code,
                                     @Field("grant_type") String grantType,
                                     @Field("client_id") String clientId,
                                     @Field("client_secret") String clientSecret,
                                     @Field("redirect_uri") String redirectUri);

    @FormUrlEncoded
    @POST("https://musicbrainz.org/oauth2/token")
    Call<AccessToken> refreshAccessToken(@Field("refresh_token") String refreshToken,
                                         @Field("grant_type") String grantType,
                                         @Field("client_id") String clientId,
                                         @Field("client_secret") String clientSecret);

    @GET("https://musicbrainz.org/oauth2/userinfo")
    Call<UserInfo> getUserInfo();


}

File: app/src/main/java/org/metabrainz/android/data/sources/api/LoginService.kt
package org.metabrainz.android.data.sources.api

import org.metabrainz.android.data.sources.api.entities.AccessToken
import org.metabrainz.android.data.sources.api.entities.userdata.UserInfo
import retrofit2.Call
import retrofit2.http.*

interface LoginService {

    @FormUrlEncoded
    @POST
    fun getAccessToken(@Url url: String?,
                       @Field("code") code: String?,
                       @Field("grant_type") grantType: String?,
                       @Field("client_id") clientId: String?,
                       @Field("client_secret") clientSecret: String?,
                       @Field("redirect_uri") redirectUri: String?): Call<AccessToken?>?

    @FormUrlEncoded
    @POST("https://musicbrainz.org/oauth2/token")
    fun refreshAccessToken(@Field("refresh_token") refreshToken: String?,
                           @Field("grant_type") grantType: String?,
                           @Field("client_id") clientId: String?,
                           @Field("client_secret") clientSecret: String?): Call<AccessToken?>?

    @get:GET("https://musicbrainz.org/oauth2/userinfo")
    val userInfo: Call<UserInfo?>?
}